### PR TITLE
Make Pauli decompose coefficient real again

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Change is a constant, but sometimes it breaks things in life. Generalizing the `pauli_decompose` in the [#4395](https://github.com/PennyLaneAI/pennylane/pull/4395) happened to be one such change. This PR adds an abstraction for the generalizes Pauli decomposition for users as a way out to help us adapt to that change (of `dtypes`). 

**Description of the Change:** 
1. `_generalized_pauli_decompose` now contains the code that was previously contained in the improved `pauli_decompose` method.
2. `pauli_decompose` now uses the above for its respective workflow while being mindful of the conventions we are meant to follow regarding `dtypes`. 

**Benefits:**
We will not be disturbing existing workflows and their expectations regarding `dtypes`.

**Possible Drawbacks:**
We might have to shift the private function to a public one at a later stage. :) 

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/pauli/conversion.py [hunk 1]
```diff
@@ -220,7 +129,212 @@ def pauli_decompose(
...
+    + -0.5 * Z(a) @ X(b)
+    + -0.5 * Z(a) @ Y(b)
+    """
+    n = int(qml.math.log2(len(H)))
+    N = 2**n
+
+    if H.shape != (N, N):
+        raise ValueError("The matrix should have shape (2**n, 2**n), for any qubit number n>=1")
+
+    if check_hermitian and not qml.math.allclose(H, qml.math.conjugate(qml.math.transpose(H))):
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:217]: might it be cheaper to just check if any coeffs have an imaginary component after computing it? this is obviously faster if it _isn't_ Hermitian, but hopefully users tend to give Hermitian matrices. As I type this I feel like I'm trying to optimize a tiny part of this so I guess it doesn't matter but just putting it out there.

File: pennylane/pauli/conversion.py [hunk 2]
```diff
@@ -220,7 +129,212 @@ def pauli_decompose(
...
+
+    if pauli:
+        return PauliSentence(
+            {
+                PauliWord({w: o for o, w in obs_n_wires}): coeff
+                for coeff, obs_n_wires in zip(coeffs, obs)
+            }
+        )
+
+    with qml.QueuingManager.stop_recording():
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:233]: nice 🤝 buuut this wasn't there before. is this considered a breaking change (that calling `pauli_decompose` doesn't queue the returned Hamiltonian)? I suppose they don't really queue on their own anyway (need to be assigned to a MeasurementProcess), so I wonder what motivated adding this.

File: pennylane/pauli/conversion.py [hunk 3]
```diff
@@ -29,114 +29,23 @@
...
 # pylint: disable=too-many-branches
-def pauli_decompose(
-    matrix,
-    hide_identity=False,
-    wire_order=None,
-    pauli=False,
-    padding=False,
-) -> Union[Hamiltonian, PauliSentence]:
+def _generalized_pauli_decompose(
+    matrix, hide_identity=False, wire_order=None, pauli=False, padding=False
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:33]: these should probably be positional args since this is an internal-only function and it's always called with all arguments specified

File: pennylane/pauli/conversion.py [hunk 4]
```diff
@@ -220,7 +129,212 @@ def pauli_decompose(
...
+        wire_order (list[Union[int, str]]): the ordered list of wires with respect
+            to which the operator is represented as a matrix.
+        pauli (bool): return a PauliSentence instance if ``True``.
+        padding (bool): makes the function compatible with rectangular matrices and square matrices
+            that are not of shape :math:`2^n\times 2^n` by padding them with zeros if ``True``.
+
+    Returns:
+        Union[Tuple[tensor_like[complex], list[~.Operation]], ~.PauliSentence]: the matrix decomposed as
+        a linear combination of Pauli operators as as a ``tuple`` consisting of an ``array`` of complex
+        coefficients and a ``list`` of corresponding Pauli terms, or :class:`~.PauliSentence` instance.
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:269]: I personally don't mind that this returns a different type from `pauli_decompose`, and it's clearly documented, but it might be a bit unexpected for users who want to try out each of the functions. I'd have expected that they work the same way except that `pauli_decompose` doesn't have a phase 😄 could you perhaps elaborate a bit on the motivation for this difference?

**Review by timmysilv — COMMENTED** [review]
thank you for the musings and the improvements! just a few concerns, but overall looks spiffy 🕴️

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -141,7 +145,7 @@ array([False, False])
 * When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
   [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)
 
-* `qml.pauli_decompose` is now differentiable and works with any non-Hermitian and non-square matrices.
+* `qml.pauli_decompose` is now fully differentiable.
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply doc/releases/changelog-dev.md:148]: Yes, it should be. Since both `pauli_decompose` and `pauli_decompose_with_phase` uses the same workflow internally.

File: pennylane/pauli/conversion.py [hunk 6]
```diff
@@ -220,7 +129,212 @@ def pauli_decompose(
...
-    # Convert to Hamiltonian and PauliSentence
+    if not pauli:
+        with qml.QueuingManager.stop_recording():
+            obs = [reduce(matmul, [op_map[o](w) for o, w in obs_term]) for obs_term in obs]
+
+    return (coeffs, obs)
+
+
+def pauli_decompose(
+    H, hide_identity=False, wire_order=None, pauli=False, check_hermitian=True
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:140]: For this method, it was initially `H` only in the previous releases. I had changed it to `matrix` in the previous PR for generalization.

File: doc/releases/changelog-dev.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply doc/releases/changelog-dev.md:148]: image

File: pennylane/pauli/conversion.py (hunk 1 above)
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:217]: I just borrowed this check from the previous implementation for `pauli_decompose` and added a keyword argument to skip this check. Even I was inclined towards adding the check you mentioned instead of this.

File: pennylane/pauli/conversion.py (hunk 2 above)
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:233]: Yes, I do think they weren't getting queued when I was doing stuff in qnode, but were getting queued when working with `qml.queuing.AnnotatedQueue()`. I thought we didn't want it so I added this. Let me know what is supposed to be the expected behaviour. 🤔

File: doc/releases/changelog-dev.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply doc/releases/changelog-dev.md:148]: and when incorporating those gradients on the next pass, will that matrix always also be hermitian?

File: pennylane/pauli/conversion.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/pauli/conversion.py:233]: not super necessary but whatever you prefer. they are "queued" in the sense that they get added to the `AnnotatedQueue` object. They are removed in the processing of that queue (eg. `QuantumScript.from_queue(q)`) unless wrapped in some MeasurementProcess (`Hamiltonian` has no "queue category" so it doesn't get added to the tape, whereas `expval(Hamiltonian)` or any observable has the measurement queue category).

File: pennylane/pauli/conversion.py [hunk 7]
```diff
@@ -29,14 +29,10 @@
...
-    matrix,
-    hide_identity=False,
-    wire_order=None,
-    pauli=False,
-    padding=False,
-) -> Union[Hamiltonian, PauliSentence]:
-    r"""Decomposes a matrix into a linear combination of Pauli operators.
+def _generalized_pauli_decompose(
+    matrix, hide_identity=False, wire_order=None, pauli=False, padding=False
+) -> Tuple[qml.math.array, list]:
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:34]: `qml.math.array` is a partial function, this would probably be better
```suggestion
) -> Tuple[qml.typing.TensorLike, list]:
```

File: pennylane/pauli/conversion.py [hunk 8]
```diff
@@ -59,83 +55,75 @@ def pauli_decompose(
...
-    + -0.5 * Y(1)
-    + -1.0 * Z(1)
-    + -1.5 * X(0)
-    + -1.0 * X(0) @ X(1)
-    + -0.5 * X(0) @ Z(1)
-    + 1.0 * Y(0) @ Y(1)
-    + -0.5 * Z(0) @ X(1)
-    + -0.5 * Z(0) @ Y(1)
+    ... [[-2, -2+1j, -2, -2], [-2-1j,  0,  0, -1], [-2,  0, -2, -1], [-2, -1, -1,  1j]])
+    >>> coeffs, obs = qml.pauli._generalized_pauli_decompose(A)
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:67]:
```suggestion
    >>> coeffs, obs = qml.pauli.conversion._generalized_pauli_decompose(A)
```

File: pennylane/pauli/conversion.py [hunk 9]
```diff
@@ -59,83 +55,75 @@ def pauli_decompose(
...
         For non-square matrices, we need to provide the ``padding=True`` keyword argument:
 
         >>> A = np.array([[-2, -2 + 1j]])
-        >>> H = qml.pauli_decompose(A, padding=True)
-        >>> print(H)
-          ((-1+0j)) [I0]
-        + ((-1+0.5j)) [X0]
-        + ((-1+0j)) [Z0]
-        + ((-0.5-1j)) [Y0]
+        >>> coeffs, obs = qml.pauli._generalized_pauli_decompose(A, padding=True)
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:110]:
```suggestion
        >>> coeffs, obs = qml.pauli.conversion._generalized_pauli_decompose(A, padding=True)
```

File: pennylane/pauli/conversion.py [hunk 10]
```diff
@@ -59,83 +55,75 @@ def pauli_decompose(
...
 
         We can also use the method within a differentiable workflow and obtain gradients:
 
         >>> A = qml.numpy.array([[-2, -2 + 1j]], requires_grad=True)
         >>> dev = qml.device("default.qubit", wires=1)
         >>> @qml.qnode(dev)
         ... def circuit(A):
-        ...    decomp = qml.pauli_decompose(A, padding=True)
-        ...    qml.RX(decomp.coeffs[2], 0)
+        ...    coeffs, _ = qml.pauli._generalized_pauli_decompose(A, padding=True)
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:122]:
```suggestion
        ...    coeffs, _ = qml.pauli.conversion._generalized_pauli_decompose(A, padding=True)
```

File: pennylane/pauli/conversion.py [hunk 11]
```diff
@@ -220,7 +208,116 @@ def pauli_decompose(
...
+        check_hermitian (bool): check if the provided matrix is Hermitian if ``True``.
+
+    Returns:
+        Union[~.Hamiltonian, ~.PauliSentence]: the matrix decomposed as a linear combination
+        of Pauli operators, either as a :class:`~.Hamiltonian` or :class:`~.PauliSentence` instance.
+
+    .. seealso::
+
+        :func:`~.pauli_decompose_with_phase` for decomposing any general matrix.
+
```
Comment (timmysilv) [inline pennylane/pauli/conversion.py:239]: whitespace might not work with suggestion but.
```suggestion
```

File: tests/pauli/test_conversion.py [hunk 12]
```diff
@@ -95,32 +207,42 @@ def test_observable_types(self, hamiltonian, hide_identity):
...
+        _, decomposed_obs = qml.pauli_decompose(hamiltonian, check_hermitian=False).terms()
         n = int(np.log2(len(hamiltonian)))
 
         tensors = filter(lambda obs: isinstance(obs, Tensor), decomposed_obs)
         assert all(len(tensor.obs) == n for tensor in tensors)
 
     @pytest.mark.parametrize("hamiltonian", test_hamiltonians)
     def test_decomposition(self, hamiltonian):
-        """Tests that pauli_decompose successfully decomposes Hamiltonians into a
+        """Tests that pauli_decompose_with_phase successfully decomposes Hamiltonians into a
```
Comment (timmysilv) [inline tests/pauli/test_conversion.py:227]:
```suggestion
        """Tests that pauli_decompose successfully decomposes Hamiltonians into a
```

File: tests/pauli/test_conversion.py [hunk 13]
```diff
@@ -188,25 +314,20 @@ def test_wire_error(self):
...
-        ):
-            qml.pauli_decompose(hamiltonian, pauli=True, wire_order=wire_order)
+            qml.pauli_decompose(hamiltonian, wire_order=wire_order, check_hermitian=False)
 
     # Multiple interfaces will be tested with math module
     @pytest.mark.all_interfaces
     @pytest.mark.parametrize("matrix", test_diff_matrix)
     def test_differentiability(self, matrix):
-        """Test differentiability for pauli_decompose"""
+        """Test differentiability for pauli_decompose_with_phase"""
```
Comment (timmysilv) [inline tests/pauli/test_conversion.py:323]:
```suggestion
        """Test differentiability for pauli_decompose"""
```

**Review by timmysilv — APPROVED** [review]
I'm cool with it 👍 😎

File: tests/pauli/test_conversion.py [hunk 14]
```diff
@@ -47,6 +48,117 @@
...
+        tensors = filter(lambda obs: isinstance(obs, Tensor), obs_list)
+
+        for tensor in tensors:
+            assert all(isinstance(o, Identity) for o in tensor.obs)
+
+    @pytest.mark.parametrize("hide_identity", [True, False])
+    @pytest.mark.parametrize("hamiltonian", test_hamiltonians)
+    def test_observable_types(self, hamiltonian, hide_identity):
+        """Tests that the Hamiltonian decomposes into a linear combination of tensors,
+        the identity matrix, and Pauli matrices."""
```
Comment (Jaybsoni) [inline tests/pauli/test_conversion.py:92]:
```suggestion
        """Tests that the Hamiltonian decomposes into a linear combination of Pauli words."""
```

File: tests/pauli/test_conversion.py [hunk 15]
```diff
@@ -47,6 +48,117 @@
...
+
+        with pytest.raises(
+            ValueError, match="number of wires 1 is not compatible with the number of qubits 2"
+        ):
+            qml.pauli_decompose(hamiltonian, wire_order=wire_order)
+
+        with pytest.raises(
+            ValueError, match="number of wires 1 is not compatible with the number of qubits 2"
+        ):
+            qml.pauli_decompose(hamiltonian, pauli=True, wire_order=wire_order)
```
Comment (Jaybsoni) [inline tests/pauli/test_conversion.py:156]:
```suggestion
        for p in [True, False]:
            with pytest.raises(
                ValueError, match="number of wires 1 is not compatible with the number of qubits 2"
            ):
                qml.pauli_decompose(hamiltonian, pauli=p, wire_order=wire_order)
```

**Review by Jaybsoni — COMMENTED** [review]
Some comments

File: pennylane/pauli/conversion.py [hunk 16]
```diff
@@ -59,83 +55,75 @@ def pauli_decompose(
             that are not of shape :math:`2^n\times 2^n` by padding them with zeros if ``True``.
 
     Returns:
-        Union[~.Hamiltonian, ~.PauliSentence]: the matrix decomposed as a linear combination
-        of Pauli operators, either as a :class:`~.Hamiltonian` or :class:`~.PauliSentence` instance.
+        Tuple[qml.math.array[complex], list]: the matrix decomposed as a linear combination of Pauli operators
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/pauli/conversion.py:58]: This would not be an operator when `pauli=True`.

**Review by Jaybsoni — APPROVED** [review]
Nice work 👍🏼
